packages/server-node/src/gmodel-lib/gmodel-operation-action-handler.ts

```diff
@@ -0,0 +1,56 @@
+/********************************************************************************
+ * Copyright (c) 2022 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { GModelRootSchema } from '@eclipse-glsp/graph';
+import { Operation, SModelRootSchema } from '@eclipse-glsp/protocol';
+import * as jsonPatch from 'fast-json-patch';
+import { inject, injectable } from 'inversify';
+import { GModelSerializer } from '../features/model/gmodel-serializer';
+import { ModelState } from '../features/model/model-state';
+import { Command } from '../features/undo-redo/command';
+import { AbstractRecordingCommand } from '../features/undo-redo/recording-command';
+import { OperationActionHandler } from '../operations/operation-action-handler';
+import { OperationHandler } from '../operations/operation-handler';
+@injectable()
+export class GModelOperationActionHandler extends OperationActionHandler {
```

Comment:
@planger  I'm not happy with they way this is currently working. The need to rebind the operation action handler just to configure the command seems not optimal to me.
Any ideas/suggestions for improvement?

Reply:
Hm, yes I see... this is indeed not ideal, also looking at [your example](https://github.com/eclipse-glsp/glsp-examples/commit/1a38dcd4c52c85d31a9b1e7189c8f954aed29485#diff-7a03dee9115e39046295b10ca9eaa8877806bff4d14816defb7521f1b409da5e), the need to rebind a source-model specific operation action handler is a bit unexpected. I'd expect only to bind a source-model-*format*-specific operation action handler (JSON, etc.), which however is generic in the sense of which actual source model (e.g. task list) is used.

Afair, I think in Java this is done differently: there we only have "source-model-format-specific" operation action handlers (e.g. EMF). Thus, if you e.g. have an EMF source model, you'd extend the abstract EMF diagram module, and that's it, because it already binds the EMF-specific operation action handler and you are expected to provide subclasses of [EMFOperationHandler](https://github.com/eclipse-glsp/glsp-server/blob/b500d905454f5cf35ff5a6c3a6b48a61245369a4/plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/EMFOperationHandler.java#L28) for your operation handlers. Thus, we delegate more to the operation handlers.

Couldn't the same approach work here too?
An `AbstractJSONOperationHandler` could require the abstract methods `getJsonObject()` and `doExecute()` to be implemented and then keep the OperationActionHandler specific only to JSON or directly to the GModel?

Of course the disadvantage is that every operation handler would need to implement `getJsonObject()`, which is redundant. In customer projects, this was typically avoided by using a common superclass for operation handlers.
So probably this is better than requiring to bind a specific operation action handler? WDYT?

Alternatively, we could be to introduce a new provider that needs to be implemented for JSON-based source models that want to benefit from JSON-patch based undo/redo support. This provider could then return the JSON object from the source model, which should eliminate the need for a specific operation action handler, right?